File: go/pactman/ein.go
package pactman

import (
	"fmt"
	"regexp"
	"strconv"
	"strings"
)

// EINLength is the number of digits in an EIN.
const EINLength = 9

// Accepted input shapes: nine digits, optionally with the conventional hyphen
// after the two-digit prefix. Surrounding whitespace is ignored.
var einPattern = regexp.MustCompile(`^\d{2}-?\d{7}$`)

// NormalizeEIN normalizes an EIN to the nine-digit form the API expects.
//
// "41-1787097" and "411787097" both normalize to "411787097".
//
// This is formatting validation only. Passing it says nothing about whether an
// organization is tax-exempt, in good standing, or eligible for anything — only
// that the value is shaped like an EIN. No IRS prefix rules are applied.
//
// A malformed value returns a *ValidationError.
func NormalizeEIN(input string) (string, error) {
	if issue, bad := einIssue(input, -1); bad {
		return "", &ValidationError{msg: issue.Message, Issues: []ValidationIssue{issue}}
	}

	return strings.Replace(strings.TrimSpace(input), "-", "", 1), nil
}

// NormalizeEINs normalizes a list of EINs, reporting every failure at once.
//
// Order is retained and duplicates are preserved; see WithDedupe for
// CheckBulk's optional deduplication. A malformed item returns a
// *ValidationError whose Issues identify each failing item by index and value.
func NormalizeEINs(inputs []string) ([]string, error) {
	var issues []ValidationIssue

	normalized := make([]string, 0, len(inputs))

	for index, input := range inputs {
		if issue, bad := einIssue(input, index); bad {
			issues = append(issues, issue)
			continue
		}

		normalized = append(normalized, strings.Replace(strings.TrimSpace(input), "-", "", 1))
	}

	if len(issues) > 0 {
		positions := make([]string, len(issues))

		for i, issue := range issues {
			positions[i] = strconv.Itoa(issue.Index)
		}

		return nil, &ValidationError{
			msg: fmt.Sprintf("%d of %d EINs are invalid (at index %s); no request was sent",
				len(issues), len(inputs), strings.Join(positions, ", ")),
			Issues: issues,
		}
	}

	return normalized, nil
}

// IsValidEIN reports whether input is shaped like an EIN.
func IsValidEIN(input string) bool {
	_, bad := einIssue(input, -1)

	return !bad
}

func einIssue(input string, index int) (ValidationIssue, bool) {
	at := ""
	if index >= 0 {
		at = fmt.Sprintf(" at index %d", index)
	}

	trimmed := strings.TrimSpace(input)

	if trimmed == "" {
		return ValidationIssue{Message: "EIN" + at + " is empty", Index: index, Value: input}, true
	}

	if !einPattern.MatchString(trimmed) {
		return ValidationIssue{
			Message: fmt.Sprintf("EIN%s must be %d digits, optionally hyphenated as XX-XXXXXXX; received %q",
				at, EINLength, trimmed),
			Index: index,
			Value: input,
		}, true
	}

	return ValidationIssue{}, false
}

File: ruby/lib/pactman/nonprofit_check_plus/ein.rb
# frozen_string_literal: true

module Pactman
  module NonprofitCheckPlus
    # EIN normalization and validation.
    #
    # Formatting validation only. Passing these checks says nothing about whether
    # an organization is tax-exempt, in good standing, or eligible for anything —
    # only that the value is shaped like an EIN. No IRS prefix rules are applied.
    module EIN
      # Number of digits in an EIN.
      LENGTH = 9

      # Accepted input shapes: nine digits, optionally with the conventional
      # hyphen after the two-digit prefix. Surrounding whitespace is ignored.
      PATTERN = /\A\d{2}-?\d{7}\z/
      private_constant :PATTERN

      # Leading and trailing whitespace, Unicode included — a no-break space
      # pasted from a document is as invisible as an ordinary one.
      SURROUNDING_WHITESPACE = /\A[[:space:]]+|[[:space:]]+\z/
      private_constant :SURROUNDING_WHITESPACE

      class << self
        # Normalizes an EIN to the nine-digit form the API expects.
        #
        # `"41-1787097"` and `"411787097"` both normalize to `"411787097"`.
        #
        # @param input [String]
        # @return [String]
        # @raise [ValidationError] if the value is not shaped like an EIN.
        def normalize(input)
          issue = issue_for(input)

          raise ValidationError.new(issue.message, [issue]) if issue

          canonical(input)
        end

        # Normalizes a collection of EINs, reporting every failure at once.
        #
        # Duplicates are preserved and order is retained; see
        # {NonprofitsResource#check_bulk} for the optional `dedupe:` behaviour.
        #
        # @param inputs [Array<String>]
        # @return [Array<String>]
        # @raise [ValidationError] if any item is not shaped like an EIN. The
        #   error's `issues` identify the failing item by index and original value.
        def normalize_all(inputs)
          issues = []
          normalized = []

          inputs.each_with_index do |input, index|
            issue = issue_for(input, index)

            if issue
              issues << issue
            else
              normalized << canonical(input)
            end
          end

          unless issues.empty?
            positions = issues.map(&:index).join(", ")

            raise ValidationError.new(
              "#{issues.size} of #{inputs.size} EINs are invalid (at index #{positions}). No request was sent.",
              issues
            )
          end

          normalized
        end

        # True when `input` is shaped like an EIN. Never raises.
        #
        # @param input [Object]
        # @return [Boolean]
        def valid?(input)
          issue_for(input).nil?
        end

        private

        def canonical(input)
          trim(input).delete("-")
        end

        def issue_for(input, index = nil)
          at = index.nil? ? "" : " at index #{index}"

          return ValidationIssue.new("EIN#{at} is required.", index, input) if input.nil?

          unless input.is_a?(String)
            return ValidationIssue.new("EIN#{at} must be a String, received #{input.class}.", index, input)
          end

          trimmed = trim(input)

          return ValidationIssue.new("EIN#{at} is empty.", index, input) if trimmed == ""

          return nil if trimmed && PATTERN.match?(trimmed)

          ValidationIssue.new(
            "EIN#{at} must be #{LENGTH} digits, optionally hyphenated as XX-XXXXXXX. " \
            "Received #{(trimmed || input).inspect}.",
            index,
            input
          )
        end

        # `nil` for a string no pattern can be run against: bytes that are not
        # valid in their declared encoding, or an encoding Regexp cannot read.
        def trim(input)
          input.gsub(SURROUNDING_WHITESPACE, "")
        rescue ArgumentError, EncodingError
          nil
        end
      end
    end
  end
end
